src/dataframe/dataframe.jl

```diff
@@ -1317,3 +1317,69 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+function _expandhelper(col, iterprod, idx)
+    for v in iterprod
+        push!(col, v[idx])
+    end
+end
+
+function expand(df::AbstractDataFrame, indexcols; error::Bool=true)
+    colind = index(df)[indexcols]
+    dummydf = similar(select(df, colind, copycols=false), 0)
+
+    if error == true
+        row_group_slots(ntuple(i -> df[!, i], ncol(df)))[1]!=nrow(df) && throw(ArgumentError("duplicate rows in input"))
+    else
+        row_group_slots(ntuple(i -> df[!, i], ncol(df)))[1]!=nrow(df) && @warn "duplicate rows in input; expand will only return unique combinations"
+    end
+
+    # Create a vect of vectors of unique values in each column
+    uniqueVals = []
+    for col in colind
+        # levels drops missing, handle the case where missing values are present
+        if any(ismissing, df[!, col])
```

Comment:
now I have noticed we have the following inconsistency: you use `levels` instead of `unique`, this means that you return all **potential** levels in a column (even if they are not present). On the other hand we add `missing` only if it is actually present (not if it is potentially present).

Is this your intention?

Reply:
It was not. Honestly speaking, I don't work with `levels` a ton, so this is just a hole in my experience/knowledge showing that I need to address :)

Reply:
@nalimilan I'd love your thoughts on this. I'd like to still be able to keep the order of categorical variables by using `levels` instead of reverting back to `unique`. The way I'm thinking is to check if the column is categorical, and if it is, calling `droplevels!` on it to ensure only categories that are in the subsetted dataset are expanded on. Is there a better approach to this?

Reply:
@nalimilan - I would keep using `levels` to retain the order of categorical variables, as @versipellis rightly commented. We have three options:
1. current functionality (retain all levels and add `missing` level only if it is present);
2. wide functionality (retain all levels and add `missing` level if it is potentially present in a categorical variable - we can detect this by examining `eltype` of column if it is categorical);
3. narrow functionality (retain only levels actually present in the column remembering their order);

Actually thinking about it I am leaning towards option 1. (current functionality) - but we just should document that we work this way. The reason is that `missing` is not a new level but rather indicates absence of information about the level from the list of of possible levels (I guess this was also your intention in CategoricalArrays.jl - right?).

Reply:
1 sounds fine. I could be convinced to take approach 3 if other implementations do that.

Reply:
OK - so let us go for 1 if @versipellis also agrees

Reply:
1 works for me as well. I'll make a note to document it as such.